File: rad/engine/02_catalog/store/definitions.go
package store

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"strconv"

	kv "github.com/Southclaws/rad/rad/engine/01_kv"
	"github.com/Southclaws/rad/rad/engine/01_kv/keyenc"
	"github.com/Southclaws/rad/rad/engine/02_catalog/model"
)

const (
	tableDefinitionPrefix = "/rad/catalog/object/table/"
	tableHeadPrefix       = "/rad/catalog/head/table/"
)

func TableDefinitionKey(id model.SchemaID, generation uint64) []byte {
	return []byte(fmt.Sprintf("%s%010d/definition/%020d", tableDefinitionPrefix, id, generation))
}

func TableDefinitionRange(id model.SchemaID) (start, end []byte) {
	prefix := []byte(fmt.Sprintf("%s%010d/definition/", tableDefinitionPrefix, id))
	return prefix, keyenc.PrefixEnd(prefix)
}

func TableHeadKey(id model.SchemaID) []byte {
	return []byte(fmt.Sprintf("%s%010d", tableHeadPrefix, id))
}

// PublishDefinitions persists the physical definitions that produced one
// canonical catalog revision. Execution pins these immutable values in memory;
// the heads are only for admission/bootstrap and are never reread by an
// already-bound plan.
func PublishDefinitions(ctx context.Context, view kv.KV, version uint64, tables []model.Table, previous model.Schema) error {
	live := make(map[model.SchemaID]bool, len(tables))
	for _, table := range tables {
		live[table.SchemaID] = true
		_, previousGeneration, hadPrevious, err := DefinitionHead(ctx, view, table.SchemaID)
		if err != nil {
			return err
		}
		raw, err := json.Marshal(table)
		if err != nil {
			return err
		}
		definitionKey := TableDefinitionKey(table.SchemaID, table.DefinitionGeneration)
		if existing, ok, err := view.Get(ctx, definitionKey); err != nil {
			return err
		} else if ok && !bytes.Equal(existing, raw) {
			return fmt.Errorf("catalog: immutable table definition %d generation %d already has different contents", table.SchemaID, table.DefinitionGeneration)
		} else if !ok {
			if err := view.Put(ctx, definitionKey, raw); err != nil {
				return err
			}
		}
		head := fmt.Sprintf("%d:%d", version, table.DefinitionGeneration)
		if err := view.Put(ctx, TableHeadKey(table.SchemaID), []byte(head)); err != nil {
			return err
		}
		if hadPrevious && previousGeneration != table.DefinitionGeneration {
			if err := QueueReclamation(ctx, view, model.Reclamation{
				ID:   TableDefinitionReclamationID(table.SchemaID, previousGeneration),
				Kind: model.ReclamationTableDefinition, RetiredCatalogVersion: version,
				TableSchemaID: table.SchemaID, DefinitionGeneration: previousGeneration,
			}); err != nil {
				return err
			}
		}
	}
	for _, table := range previous.Tables {
		if live[table.ID] {
			continue
		}
		_, previousGeneration, ok, err := DefinitionHead(ctx, view, table.ID)
		if err != nil {
			return err
		}
		if ok {
			if err := QueueReclamation(ctx, view, model.Reclamation{
				ID:   TableDefinitionReclamationID(table.ID, previousGeneration),
				Kind: model.ReclamationTableDefinition, RetiredCatalogVersion: version,
				TableSchemaID: table.ID, DefinitionGeneration: previousGeneration,
			}); err != nil {
				return err
			}
		}
		if err := view.Delete(ctx, TableHeadKey(table.ID)); err != nil {
			return err
		}
	}
	return nil
}

func DefinitionHead(ctx context.Context, view kv.KV, id model.SchemaID) (catalogVersion, definitionGeneration uint64, ok bool, err error) {
	raw, ok, err := view.Get(ctx, TableHeadKey(id))
	if err != nil || !ok {
		return 0, 0, ok, err
	}
	var split int
	for split < len(raw) && raw[split] != ':' {
		split++
	}
	if split == len(raw) {
		return 0, 0, false, fmt.Errorf("catalog: corrupt table definition head %q", raw)
	}
	catalogVersion, err = strconv.ParseUint(string(raw[:split]), 10, 64)
	if err != nil {
		return 0, 0, false, fmt.Errorf("catalog: corrupt table definition catalog version %q: %w", raw, err)
	}
	definitionGeneration, err = strconv.ParseUint(string(raw[split+1:]), 10, 64)
	if err != nil {
		return 0, 0, false, fmt.Errorf("catalog: corrupt table definition generation %q: %w", raw, err)
	}
	return catalogVersion, definitionGeneration, true, nil
}

File: src/engine/02_catalog/store/definitions.rs
use std::collections::HashSet;

use bytes::Bytes;

use crate::engine::catalog::identity::{CatalogVersion, DefinitionGeneration, SchemaId};
use crate::engine::catalog::model::{Reclamation, ReclamationKind, Schema, Table, Timestamp};
use crate::engine::catalog::{Error, ErrorKind, Result};
use crate::engine::kv::KvView;

use super::durable_json::encode;
use super::{map_kv, parse_u64, prefix_bounds, queue_reclamation, table_definition_reclamation_id};

const TABLE_DEFINITION_PREFIX: &str = "/rad/catalog/object/table/";
const TABLE_HEAD_PREFIX: &str = "/rad/catalog/head/table/";

pub fn table_definition_key(id: SchemaId, generation: DefinitionGeneration) -> Vec<u8> {
    format!(
        "{TABLE_DEFINITION_PREFIX}{:010}/definition/{:020}",
        id.get(),
        generation.get()
    )
    .into_bytes()
}

pub fn table_definition_range(id: SchemaId) -> (Vec<u8>, Vec<u8>) {
    let start = format!("{TABLE_DEFINITION_PREFIX}{:010}/definition/", id.get()).into_bytes();
    prefix_bounds(start)
}

pub fn table_head_key(id: SchemaId) -> Vec<u8> {
    format!("{TABLE_HEAD_PREFIX}{:010}", id.get()).into_bytes()
}

pub async fn publish_definitions<V: KvView + ?Sized>(
    view: &mut V,
    version: CatalogVersion,
    tables: &[Table],
    previous: &Schema,
    now: Timestamp,
) -> Result<()> {
    let mut live = HashSet::with_capacity(tables.len());
    for table in tables {
        live.insert(table.schema_id);
        let previous_head = definition_head(view, table.schema_id).await?;
        let raw = encode("table definition", &table.schema_id.to_string(), table)?;
        let definition_key = table_definition_key(table.schema_id, table.definition_generation);
        if let Some(existing) = view.get(&definition_key).await.map_err(map_kv)? {
            if existing.as_ref() != raw.as_slice() {
                return Err(Error::message(
                    ErrorKind::Conflict,
                    format!(
                        "catalog: immutable table definition {} generation {} already has different contents",
                        table.schema_id, table.definition_generation
                    ),
                ));
            }
        } else {
            view.put(Bytes::from(definition_key), Bytes::from(raw))
                .await
                .map_err(map_kv)?;
        }
        view.put(
            Bytes::from(table_head_key(table.schema_id)),
            Bytes::from(format!("{version}:{}", table.definition_generation)),
        )
        .await
        .map_err(map_kv)?;

        if let Some((_, previous_generation)) = previous_head
            && previous_generation != table.definition_generation
        {
            let mut reclamation = Reclamation::pending(
                table_definition_reclamation_id(table.schema_id, previous_generation),
                ReclamationKind::TableDefinition,
                version,
                now,
            );
            reclamation.table_schema_id = Some(table.schema_id);
            reclamation.definition_generation = previous_generation;
            queue_reclamation(view, reclamation, now).await?;
        }
    }

    for table in &previous.tables {
        if live.contains(&table.id) {
            continue;
        }
        if let Some((_, previous_generation)) = definition_head(view, table.id).await? {
            let mut reclamation = Reclamation::pending(
                table_definition_reclamation_id(table.id, previous_generation),
                ReclamationKind::TableDefinition,
                version,
                now,
            );
            reclamation.table_schema_id = Some(table.id);
            reclamation.definition_generation = previous_generation;
            queue_reclamation(view, reclamation, now).await?;
        }
        view.delete(&table_head_key(table.id))
            .await
            .map_err(map_kv)?;
    }
    Ok(())
}

pub async fn definition_head<V: KvView + ?Sized>(
    view: &mut V,
    id: SchemaId,
) -> Result<Option<(CatalogVersion, DefinitionGeneration)>> {
    let Some(raw) = view.get(&table_head_key(id)).await.map_err(map_kv)? else {
        return Ok(None);
    };
    let value = std::str::from_utf8(&raw).map_err(|error| {
        Error::source(
            ErrorKind::CatalogCorrupt,
            format!("catalog: corrupt table definition head {raw:?}"),
            error,
        )
    })?;
    let Some((version, generation)) = value.split_once(':') else {
        return Err(Error::message(
            ErrorKind::CatalogCorrupt,
            format!("catalog: corrupt table definition head {raw:?}"),
        ));
    };
    let version = parse_u64("definition version", None, version.as_bytes())?;
    let generation = parse_u64("definition generation", None, generation.as_bytes())?;
    Ok(Some((version.into(), generation.into())))
}

#[cfg(test)]
mod tests {
    use crate::engine::catalog::identity::{ExistenceGeneration, TableId, WriteProtocolGeneration};
    use crate::engine::catalog::store::list_reclamations;
    use crate::engine::kv::TransactionalKv;
    use crate::engine::kv::slatedb;

    use super::*;

    fn table(generation: u64) -> Table {
        Table {
            id: TableId::from("t1"),
            schema_id: SchemaId::new(1).unwrap(),
            name: "users".into(),
            definition_generation: generation.into(),
            existence_generation: ExistenceGeneration::from(1),
            write_protocol_generation: WriteProtocolGeneration::from(1),
            columns: Vec::new(),
            primary_key: Vec::new(),
            indexes: Vec::new(),
            foreign_keys: Vec::new(),
            constraints: Vec::new(),
        }
    }

    #[tokio::test]
    async fn definitions_are_immutable_and_replaced_heads_queue_reclamation() {
        let mut database = slatedb::Store::memory("catalog-definitions").await.unwrap();
        publish_definitions(
            &mut database,
            1.into(),
            &[table(1)],
            &Schema::default(),
            Timestamp::test_value(),
        )
        .await
        .unwrap();
        let previous = Schema::from_physical(&[table(1)]).unwrap();
        publish_definitions(
            &mut database,
            2.into(),
            &[table(2)],
            &previous,
            Timestamp::test_value(),
        )
        .await
        .unwrap();
        assert_eq!(
            definition_head(&mut database, SchemaId::new(1).unwrap())
                .await
                .unwrap(),
            Some((2.into(), 2.into()))
        );
        assert_eq!(list_reclamations(&mut database).await.unwrap().len(), 1);
        database.close().await.unwrap();
    }
}
